### PR TITLE
Fixed bug enable and disable module

### PR DESCRIPTION
**Pull request by christianverardi**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | If you have installed a module that has some overrides and you are in a multishop configuration and you try to disable the module on one shop because you don't need it in that shop you will lose the overrides causing the malfunction of the module on the other shop. This solution makes the Module Class to uninstall/install overrides if the disable/enable action is called for all shops. It should be a good practice of the programmer to check in the overridden methods if the module is enabled or not. 
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |
| How to test?  | Go to the Admin Modules page, select a shop in case of multishop configuration and disable or enable a module that has some overrides.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8187)

### CODE REVIEW

**Comment by xBorderie** [comment]
Thank you for remaking #8182!
Now let's see what the Core team thinks :)

**Comment by LittleBigDev** [comment]
Hello @christianverardi, could you please fix your code so unit tests don't fail anymore ?

**Comment by xBorderie** [comment]
@christianverardi Error seems to lie here: https://travis-ci.org/PrestaShop/PrestaShop/jobs/258042996#L877

File: classes/module/Module.php
Comment (eternoendless) [inline classes/module/Module.php]: It looks like this will prevent overrides from being installed when you enable a module on a single shop.

Comment (christianverardi) [reply classes/module/Module.php]: Hi @eternoendless I know, but the default behavior is not correct as well because if you disable the module only on one shop you will lose the overrides for the other shop.
I'm trying to edit my solution in order to take the count of the existing shops and compare it with the getContextListShopID. If they matches we can uninstall the overrides otherwise they should not be uninstalled.

File: classes/module/Module.php
Comment (LittleBigDev) [inline classes/module/Module.php]: `$force_all` is parameter of `ModuleCore::enable()`. It should not be modified by it.
Determining which shops are concerned about module enabling and disabling should not be done *inside* `enable()` and `disable()`.
Moreover, you are trying to deflect `$force_all` usage - which concerns enabling/disabling modules - in order to change overrides install/uninstall behaviour.
~I suggest you to have a look at the way overrides are uninstalled and find a way to fix the behaviour there.~
Sadly this is impossible. Overrides are installed / uninstalled for all shops. They can't be ignored for specific shops, and "active" for other ones.

As enabling/disabling a module only affect its behavior, and has nothing to do with overrides, there is no solution to your current issue.

File: classes/module/Module.php
Comment (LittleBigDev) [inline classes/module/Module.php]: `$force_all` is parameter of `ModuleCore::enable()`. It should not be modified by it.
Determining which shops are concerned about module enabling and disabling should not be done *inside* `enable()` and `disable()`.
Moreover, you are trying to deflect `$force_all` usage - which concerns enabling/disabling modules - in order to change overrides install/uninstall behaviour.
~I suggest you to have a look at the way overrides are uninstalled and find a way to fix the behaviour there.~
Sadly this is impossible. Overrides are installed / uninstalled for all shops. They can't be ignored for specific shops, and "active" for other ones.

As enabling/disabling a module only affect its behavior, and has nothing to do with overrides, there is no solution to your current issue.

File: classes/module/Module.php
Comment (LittleBigDev) [inline classes/module/Module.php:762]: `$force_all` is parameter of `ModuleCore::enable()`. It should not be modified by it.
Determining which shops are concerned about module enabling and disabling should not be done *inside* `enable()` and `disable()`.
Moreover, you are trying to deflect `$force_all` usage - which concerns enabling/disabling modules - in order to change overrides install/uninstall behaviour.
~I suggest you to have a look at the way overrides are uninstalled and find a way to fix the behaviour there.~
Sadly this is impossible. Overrides are installed / uninstalled for all shops. They can't be ignored for specific shops, and "active" for other ones.

As enabling/disabling a module only affect its behavior, and has nothing to do with overrides, there is no solution to your current issue.

**Review by LittleBigDev — CHANGES_REQUESTED** [review]

File: classes/module/Module.php
Comment (christianverardi) [reply classes/module/Module.php:762]: @LittleBigDev you are right but your sentence is not

> As enabling/disabling a module only affect its behavior, and has nothing to do with overrides, there is no solution to your current issue.

 because enabling/disabling a module will install/uninstall its overrides [https://github.com/PrestaShop/PrestaShop/blob/develop/classes/module/Module.php#L867](https://github.com/PrestaShop/PrestaShop/blob/develop/classes/module/Module.php#L867)
This is not correct because it will cause the malfunction of the module on the other shop (on which the module is enabled).

I think that lines 761-770 and 866-868 should be totally removed.

Comment (christianverardi) [reply classes/module/Module.php:762]: @LittleBigDev these are the wrong commits I think:
[https://github.com/PrestaShop/PrestaShop/commit/f1fdee320c6713b67ff37b2eb808bf761e9859fd](https://github.com/PrestaShop/PrestaShop/commit/f1fdee320c6713b67ff37b2eb808bf761e9859fd)
[https://github.com/PrestaShop/PrestaShop/commit/7d6c52da604c677ae4e454329313904e166fbe5f](https://github.com/PrestaShop/PrestaShop/commit/7d6c52da604c677ae4e454329313904e166fbe5f)
As you said enabling/disabling a module should only affect its behavior and not its overrides unless it is enabled or disabled on all shops.

Comment (LittleBigDev) [reply classes/module/Module.php:762]: Sorry Christian, I think I misspoke.
I totally agree with you, you are pointing at a known issue. For now, we cannot merge this PR because it would introduce one or more regressions.
Core team has planned to discuss about the issue and fix it.

Comment (christianverardi) [reply classes/module/Module.php:762]: ok @LittleBigDev thank you for the update!

**Comment by matks** [comment]
Hi, [following our review guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/) "Pull Requests may be closed after 30 days of inactivity following a request for modifications".

Since this PR did not get any activity since 2017, I close it.

Please note this does not mean it's a final "won't merge" 😉 if you come back we can reopen it and update it

Thank you for the contribution

**Pull request closed without merging**
